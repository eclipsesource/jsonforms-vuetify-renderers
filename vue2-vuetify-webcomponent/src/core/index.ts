export * from './dynamicDefaults';
export * from './i18n';
export * from './json-refs';
export * from './keywords';
export * from './transform';
export * from './types';
export * from './validate';
