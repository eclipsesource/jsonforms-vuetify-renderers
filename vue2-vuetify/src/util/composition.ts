import {
  composePaths,
  computeLabel,
  getFirstPrimitiveProp,
  isDescriptionHidden,
  JsonFormsSubStates,
  Resolve,
} from '@jsonforms/core';
import cloneDeep from 'lodash/cloneDeep';
import merge from 'lodash/merge';
import { useStyles } from '../styles';
import { computed, ComputedRef, inject, ref } from '../vue';
import { Ajv } from 'ajv';

const useControlAppliedOptions = <I extends { control: any }>(input: I) => {
  return computed(() =>
    merge(
      {},
      cloneDeep(input.control.value.config),
      cloneDeep(input.control.value.uischema.options)
    )
  );
};

const useComputedLabel = <I extends { control: any }>(
  input: I,
  appliedOptions: ComputedRef<any>
) => {
  return computed((): string => {
    return computeLabel(
      input.control.value.label,
      input.control.value.required,
      !!appliedOptions.value?.hideRequiredAsterisk
    );
  });
};

/**
 * Adds styles, isFocused, appliedOptions and onChange
 */
export const useVuetifyControl = <
  I extends { control: any; handleChange: any }
>(
  input: I,
  adaptValue: (target: any) => any = (v) => v
) => {
  const appliedOptions = useControlAppliedOptions(input);

  const isFocused = ref(false);
  const onChange = (value: any) => {
    input.handleChange(input.control.value.path, adaptValue(value));
  };

  const persistentHint = (): boolean => {
    return !isDescriptionHidden(
      input.control.value.visible,
      input.control.value.description,
      isFocused.value,
      !!appliedOptions.value?.showUnfocusedDescription
    );
  };

  const computedLabel = useComputedLabel(input, appliedOptions);

  const controlWrapper = computed(() => {
    const { id, description, errors, label, visible, required } =
      input.control.value;
    return { id, description, errors, label, visible, required };
  });

  const styles = useStyles(input.control.value.uischema);

  return {
    ...input,
    styles,
    isFocused,
    appliedOptions,
    controlWrapper,
    onChange,
    persistentHint,
    computedLabel,
  };
};

/**
 * Adds styles and appliedOptions
 */
export const useVuetifyLayout = <I extends { layout: any }>(input: I) => {
  const appliedOptions = computed(() =>
    merge(
      {},
      cloneDeep(input.layout.value.config),
      cloneDeep(input.layout.value.uischema.options)
    )
  );
  return {
    ...input,
    styles: useStyles(input.layout.value.uischema),
    appliedOptions,
  };
};

/**
 * Adds styles, appliedOptions and childUiSchema
 */
export const useVuetifyArrayControl = <I extends { control: any }>(
  input: I
) => {
  const appliedOptions = useControlAppliedOptions(input);

  const computedLabel = useComputedLabel(input, appliedOptions);

  const childLabelForIndex = (index: number) => {
    const childLabelProp =
      input.control.value.uischema.options?.childLabelProp ??
      getFirstPrimitiveProp(input.control.value.schema);
    if (!childLabelProp) {
      return `${index}`;
    }
    const labelValue = Resolve.data(
      input.control.value.data,
      composePaths(`${index}`, childLabelProp)
    );
    if (
      labelValue === undefined ||
      labelValue === null ||
      Number.isNaN(labelValue)
    ) {
      return '';
    }
    return `${labelValue}`;
  };
  return {
    ...input,
    styles: useStyles(input.control.value.uischema),
    appliedOptions,
    childLabelForIndex,
    computedLabel,
  };
};

/**
 * Adds styles and appliedOptions
 */
export const useVuetifyBasicControl = <I extends { control: any }>(
  input: I
) => {
  const appliedOptions = useControlAppliedOptions(input);

  return {
    ...input,
    styles: useStyles(input.control.value.uischema),
    appliedOptions,
  };
};

/**
 * Extracts Ajv from JSON Forms
 */
export const useAjv = () => {
  const jsonforms = inject<JsonFormsSubStates>('jsonforms');

  if (!jsonforms) {
    throw new Error(
      "'jsonforms' couldn't be injected. Are you within JSON Forms?"
    );
  }

  // should always exist
  return jsonforms.core?.ajv as Ajv;
};
