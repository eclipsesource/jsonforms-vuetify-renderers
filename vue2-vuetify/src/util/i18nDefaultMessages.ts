export const i18nDefaultMessages = {
  arraylayout: {
    add: 'Add',
    delete: 'Delete',
    moveUp: 'Move Up',
    moveDown: 'Move Down',
    dialogTitle: 'Delete {{ element }}?',
    dialogText: 'The element will be deleted.',
    dialogConfirm: 'Delete',
    dialogCancel: 'Cancel',
  },
};
