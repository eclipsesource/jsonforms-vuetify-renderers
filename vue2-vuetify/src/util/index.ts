export * from './options';
export * from './composition';
export * from './validator';
export * from './datejs';
export * from './i18nDefaultMessages';
