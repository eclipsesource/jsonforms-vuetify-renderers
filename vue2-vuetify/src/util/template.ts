import templateFn from 'lodash/template';
import merge from 'lodash/merge';
import { TemplateExecutor, TemplateOptions } from 'lodash';

const interpolate = /{{([\s\S]+?)}}/g;

export const template = (
  templateString?: string,
  options?: TemplateOptions
): TemplateExecutor => {
  return templateFn(
    templateString,
    merge({ interpolate: interpolate }, options)
  );
};
