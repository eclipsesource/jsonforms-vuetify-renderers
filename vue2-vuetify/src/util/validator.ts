import { Options } from 'ajv';
import { createAjv as createAjvCore } from '@jsonforms/core';
import Ajv from 'ajv';

export const createAjv = (options?: Options): Ajv => {
  const ajv = createAjvCore(options);
  ajv.addFormat('password', (_) => true);
  return ajv;
};
