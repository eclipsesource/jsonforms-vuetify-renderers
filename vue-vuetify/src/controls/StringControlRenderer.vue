<template>
  <p class="text-blue">live3</p>
  <p>{{touched}}</p>
  <p>{{filteredErrors}}</p>
  <control-wrapper
    v-bind="controlWrapper"
    :styles="styles"
    :isFocused="isFocused"
    :appliedOptions="appliedOptions"
  >
    <v-hover v-slot="{ isHovering }">
      <v-combobox
        v-if="suggestions !== undefined"
        v-disabled-icon-focus
        :id="control.id + '-input'"
        :class="styles.control.input"
        :disabled="!control.enabled"
        :autofocus="appliedOptions.focus"
        :placeholder="appliedOptions.placeholder"
        :label="computedLabel"
        :hint="control.description"
        :persistent-hint="persistentHint()"
        :required="control.required"
        :error-messages="control.errors"
        :maxlength="
          appliedOptions.restrict ? control.schema.maxLength : undefined
        "
        :counter="
          control.schema.maxLength !== undefined
            ? control.schema.maxLength
            : undefined
        "
        :clearable="isHovering"
        :model-value="control.data"
        :items="suggestions"
        hide-no-data
        v-bind="vuetifyProps('v-combobox')"
        @update:model-value="onChange"
        @focus="isFocused = true"
        @blur="handleBlur"
      />
      <v-text-field
        v-else
        :id="control.id + '-input'"
        :class="styles.control.input"
        :disabled="!control.enabled"
        :autofocus="appliedOptions.focus"
        :placeholder="appliedOptions.placeholder"
        :label="computedLabel"
        :hint="control.description"
        :persistent-hint="persistentHint()"
        :required="control.required"
        :error-messages="control.errors"
        :model-value="control.data"
        :maxlength="
          appliedOptions.restrict ? control.schema.maxLength : undefined
        "
        :counter="
          control.schema.maxLength !== undefined
            ? control.schema.maxLength
            : undefined
        "
        :clearable="isHovering"
        v-bind="vuetifyProps('v-text-field')"
        @update:model-value="onChange"
        @focus="isFocused = true"
        @blur="handleBlur"
      />
    </v-hover>
  </control-wrapper>
</template>

<script lang="ts">
import {
  ControlElement,
  JsonFormsRendererRegistryEntry,
  rankWith,
  isStringControl,
} from '@jsonforms/core';
import {defineComponent} from 'vue';
import {
  rendererProps,
  useJsonFormsControl,
  RendererProps,
} from '@jsonforms/vue';
import { default as ControlWrapper } from './ControlWrapper.vue';
import { useVuetifyControl } from '../util';
import { VHover, VTextField, VCombobox } from 'vuetify/components';
import { DisabledIconFocus } from './directives';
import isArray from 'lodash/isArray';
import every from 'lodash/every';
import isString from 'lodash/isString';
import { useBlurHandler } from "../util";

const controlRenderer = defineComponent({
  name: 'string-control-renderer',
  components: {
    ControlWrapper,
    VHover,
    VTextField,
    VCombobox,
  },
  directives: {
    DisabledIconFocus,
  },
  props: {
    ...rendererProps<ControlElement>(),
  },
  setup(props: RendererProps<ControlElement>) {
    const vControl = useVuetifyControl(
      useJsonFormsControl(props),
      (value) => value || undefined,
      300
    );

    const { handleBlur } = useBlurHandler(vControl);

    return {
      ...vControl,
      handleBlur
    }
  },
  computed: {
    suggestions(): string[] | undefined {
      const suggestions = this.control.uischema.options?.suggestion;

      if (
        suggestions === undefined ||
        !isArray(suggestions) ||
        !every(suggestions, isString)
      ) {
        // check for incorrect data
        return undefined;
      }
      return suggestions;
    },
  },
});

export default controlRenderer;

export const entry: JsonFormsRendererRegistryEntry = {
  renderer: controlRenderer,
  tester: rankWith(1, isStringControl),
};
</script>
