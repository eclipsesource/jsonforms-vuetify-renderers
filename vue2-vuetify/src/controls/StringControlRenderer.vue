<template>
  <control-wrapper
    v-bind="controlWrapper"
    :styles="styles"
    :isFocused="isFocused"
    :appliedOptions="appliedOptions"
  >
    <v-hover v-slot="{ hover }">
      <v-combobox
        v-if="suggestions !== undefined"
        v-disabled-icon-focus
        :id="control.id + '-input'"
        :class="styles.control.input"
        :disabled="!control.enabled"
        :autofocus="appliedOptions.focus"
        :placeholder="appliedOptions.placeholder"
        :label="computedLabel"
        :hint="control.description"
        :persistent-hint="persistentHint()"
        :required="control.required"
        :error-messages="control.errors"
        :maxlength="
          appliedOptions.restrict ? control.schema.maxLength : undefined
        "
        :counter="
          control.schema.maxLength !== undefined
            ? control.schema.maxLength
            : undefined
        "
        :clearable="hover"
        :value="control.data"
        :items="suggestions"
        hide-no-data
        @change="onChange"
        @focus="isFocused = true"
        @blur="isFocused = false"
      />
      <v-text-field
        v-else
        v-disabled-icon-focus
        :id="control.id + '-input'"
        :class="styles.control.input"
        :disabled="!control.enabled"
        :autofocus="appliedOptions.focus"
        :placeholder="appliedOptions.placeholder"
        :label="computedLabel"
        :hint="control.description"
        :persistent-hint="persistentHint()"
        :required="control.required"
        :error-messages="control.errors"
        :value="control.data"
        :maxlength="
          appliedOptions.restrict ? control.schema.maxLength : undefined
        "
        :counter="
          control.schema.maxLength !== undefined
            ? control.schema.maxLength
            : undefined
        "
        :clearable="hover"
        @change="onChange"
        @focus="isFocused = true"
        @blur="isFocused = false"
      />
    </v-hover>
  </control-wrapper>
</template>

<script lang="ts">
import {
  ControlElement,
  JsonFormsRendererRegistryEntry,
  rankWith,
  isStringControl,
} from '@jsonforms/core';
import { defineComponent } from '../vue';
import {
  rendererProps,
  useJsonFormsControl,
  RendererProps,
} from '@jsonforms/vue2';
import { default as ControlWrapper } from './ControlWrapper.vue';
import { useVuetifyControl } from '../util';
import { VHover, VTextField, VCombobox } from 'vuetify/lib';
import { DisabledIconFocus } from './directives';
import isArray from 'lodash/isArray';
import every from 'lodash/every';
import isString from 'lodash/isString';

const controlRenderer = defineComponent({
  name: 'string-control-renderer',
  components: {
    ControlWrapper,
    VHover,
    VTextField,
    VCombobox,
  },
  directives: {
    DisabledIconFocus,
  },
  props: {
    ...rendererProps<ControlElement>(),
  },
  setup(props: RendererProps<ControlElement>) {
    return useVuetifyControl(
      useJsonFormsControl(props),
      (value) => value || undefined
    );
  },
  computed: {
    suggestions(): string[] | undefined {
      const suggestions = this.control.uischema.options?.suggestion;

      if (
        suggestions === undefined ||
        !isArray(suggestions) ||
        !every(suggestions, isString)
      ) {
        // check for incorrect data
        return undefined;
      }
      return suggestions;
    },
  },
});

export default controlRenderer;

export const entry: JsonFormsRendererRegistryEntry = {
  renderer: controlRenderer,
  tester: rankWith(1, isStringControl),
};
</script>
