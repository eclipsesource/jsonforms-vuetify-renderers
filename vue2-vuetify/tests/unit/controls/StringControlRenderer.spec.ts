import { clearAllIds } from '@jsonforms/core';
import { Wrapper } from '@vue/test-utils';
import StringControlRenderer, {
  entry as stringControlRendererEntry,
} from '../../../src/controls/StringControlRenderer.vue';
import { wait } from '../../../tests';
import { mountJsonForms } from '../util';

describe('StringControlRenderer.vue', () => {
  const renderers = [stringControlRendererEntry];

  const data = 'a';
  const schema = {
    type: 'string',
    title: 'My String',
  };
  const uischema = {
    type: 'Control',
    scope: '#',
    options: {
      placeholder: 'string placeholder',
    },
  };

  let wrapper: Wrapper<any, Element>;

  beforeEach(() => {
    // clear all ids to guarantee that the snapshots will always be generated with the same ids
    clearAllIds();
    wrapper = mountJsonForms(data, schema, renderers, uischema);
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('check if child StringControlRenderer exists', () => {
    expect(wrapper.getComponent(StringControlRenderer));
  });

  it('renders a string input', () => {
    const input = wrapper.find('input[type="text"]');
    expect(input.exists()).toEqual(true);
  });

  it('renders title as label', () => {
    expect(wrapper.find('label').text()).toEqual('My String');
  });

  it('emits a data change', async () => {
    const input = wrapper.find('input[type="text"]');
    await input.setValue('b');
    // 300 ms debounceWait
    await wait(300);
    expect(wrapper.vm.$data.data).toEqual('b');
  });

  it('should have a placeholder', async () => {
    const input = wrapper.find('input[type="text"]');
    await input.trigger('focus');
    const placeholder = input.attributes('placeholder');
    expect(placeholder).toEqual('string placeholder');
  });

  it('should render component and match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe('StringControlRenderer.vue with suggestion', () => {
  const renderers = [stringControlRendererEntry];

  const data = '';

  const schema = {
    type: 'string',
    title: 'Select Continent',
  };
  const uischema = {
    type: 'Control',
    scope: '#',
    options: {
      placeholder: 'Enter your continent',
      suggestion: [
        'Asia',
        'Africa',
        'Europe',
        'North America',
        'South America',
        'Australia/Oceania',
        'Antarctica',
      ],
    },
  };

  let wrapper: Wrapper<any, Element>;

  beforeEach(() => {
    // clear all ids to guarantee that the snapshots will always be generated with the same ids
    clearAllIds();
    wrapper = mountJsonForms(data, schema, renderers, uischema);
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('check if child StringControlRenderer exists', () => {
    expect(wrapper.getComponent(StringControlRenderer));
  });

  it('renders a string input', () => {
    const input = wrapper.find('input[type="text"]');
    expect(input.exists()).toEqual(true);
    expect(input.attributes('type')).toEqual('text');
  });

  it('renders title as label', () => {
    expect(wrapper.find('label').text()).toEqual('Select Continent');
  });

  it('emits a data change', async () => {
    const input = wrapper.find('input[type="text"]');
    await input.setValue('Europe');
    // select the element
    await input.trigger('keydown.enter');
    // 300 ms debounceWait
    await wait(300);
    expect(wrapper.vm.$data.data).toEqual('Europe');
  });

  it('should have a placeholder', async () => {
    const input = wrapper.find('input[type="text"]');
    // select the input so the placeholder is generated
    await input.trigger('click');

    const placeholder = input.attributes('placeholder');
    expect(placeholder).toEqual('Enter your continent');
  });

  it('should render component and match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render component and match snapshot when clicked', async () => {
    const input = wrapper.find('input[type="text"]');
    await input.trigger('click');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
