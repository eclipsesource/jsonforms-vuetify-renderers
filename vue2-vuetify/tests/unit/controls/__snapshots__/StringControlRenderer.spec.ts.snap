// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StringControlRenderer.vue should render component and match snapshot 1`] = `
<div data-app="true"
     class="v-application v-application--is-ltr theme--light"
     id="app"
>
  <div class="v-application--wrap">
    <div id="#"
         class="control"
         appliedoptions="[object Object]"
         errors
         label="My String"
    >
      <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--placeholder input">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot">
              <label for="#-input"
                     class="v-label v-label--active theme--light"
                     style="left: 0px; position: absolute;"
              >
                My String
              </label>
              <input disabled-icon-focus
                     id="#-input"
                     type="text"
              >
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div"
                                     name="message-transition"
                                     class="v-messages__wrapper"
              >
              </transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`StringControlRenderer.vue with suggestion should render component and match snapshot 1`] = `
<div data-app="true"
     class="v-application v-application--is-ltr theme--light"
     id="app"
>
  <div class="v-application--wrap">
    <div id="#"
         class="control"
         appliedoptions="[object Object]"
         errors
         label="Select Continent"
    >
      <div class="v-input theme--light v-text-field v-text-field--placeholder v-select v-autocomplete input">
        <div class="v-input__control">
          <div role="combobox"
               aria-haspopup="listbox"
               aria-expanded="false"
               aria-owns="list-150"
               class="v-input__slot"
          >
            <div class="v-select__slot">
              <label for="#-input"
                     class="v-label theme--light"
                     style="left: 0px; position: absolute;"
              >
                Select Continent
              </label>
              <input disabled-icon-focus
                     id="#-input"
                     type="text"
                     autocomplete="off"
              >
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append">
                  <i aria-hidden="true"
                     class="v-icon notranslate mdi mdi-menu-down theme--light"
                  >
                  </i>
                </div>
              </div>
              <input type="hidden">
            </div>
            <div class="v-menu">
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div"
                                     name="message-transition"
                                     class="v-messages__wrapper"
              >
              </transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`StringControlRenderer.vue with suggestion should render component and match snapshot when clicked 1`] = `
<div data-app="true"
     class="v-application v-application--is-ltr theme--light"
     id="app"
>
  <div class="v-application--wrap">
    <div id="#"
         class="control"
         appliedoptions="[object Object]"
         errors
         label="Select Continent"
         isfocused="true"
    >
      <div class="v-input v-input--is-focused theme--light v-text-field v-text-field--placeholder v-select v-select--is-menu-active v-autocomplete primary--text input">
        <div class="v-input__control">
          <div role="combobox"
               aria-haspopup="listbox"
               aria-expanded="true"
               aria-owns="list-163"
               class="v-input__slot"
          >
            <div class="v-select__slot">
              <label for="#-input"
                     class="v-label v-label--active theme--light primary--text"
                     style="left: 0px; position: absolute;"
              >
                Select Continent
              </label>
              <input disabled-icon-focus
                     id="#-input"
                     type="text"
                     autocomplete="off"
                     placeholder="Enter your continent"
              >
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append">
                  <i aria-hidden="true"
                     class="v-icon notranslate mdi mdi-menu-down theme--light primary--text"
                  >
                  </i>
                </div>
              </div>
              <input type="hidden">
            </div>
            <div class="v-menu">
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light primary--text">
              <transition-group-stub tag="div"
                                     name="message-transition"
                                     class="v-messages__wrapper"
              >
              </transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-menu__content theme--light menuable__content__active v-autocomplete__content"
       style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 8; display: none;"
  >
    <div role="listbox"
         tabindex="-1"
         class="v-list v-select-list v-sheet theme--light theme--light"
         id="list-163"
    >
      <div tabindex="0"
           aria-selected="false"
           id="list-item-169-0"
           role="option"
           class="v-list-item v-list-item--link theme--light"
      >
        <div class="v-list-item__content">
          <div class="v-list-item__title">
            Asia
          </div>
        </div>
      </div>
      <div tabindex="0"
           aria-selected="false"
           id="list-item-169-1"
           role="option"
           class="v-list-item v-list-item--link theme--light"
      >
        <div class="v-list-item__content">
          <div class="v-list-item__title">
            Africa
          </div>
        </div>
      </div>
      <div tabindex="0"
           aria-selected="false"
           id="list-item-169-2"
           role="option"
           class="v-list-item v-list-item--link theme--light"
      >
        <div class="v-list-item__content">
          <div class="v-list-item__title">
            Europe
          </div>
        </div>
      </div>
      <div tabindex="0"
           aria-selected="false"
           id="list-item-169-3"
           role="option"
           class="v-list-item v-list-item--link theme--light"
      >
        <div class="v-list-item__content">
          <div class="v-list-item__title">
            North America
          </div>
        </div>
      </div>
      <div tabindex="0"
           aria-selected="false"
           id="list-item-169-4"
           role="option"
           class="v-list-item v-list-item--link theme--light"
      >
        <div class="v-list-item__content">
          <div class="v-list-item__title">
            South America
          </div>
        </div>
      </div>
      <div tabindex="0"
           aria-selected="false"
           id="list-item-169-5"
           role="option"
           class="v-list-item v-list-item--link theme--light"
      >
        <div class="v-list-item__content">
          <div class="v-list-item__title">
            Australia/Oceania
          </div>
        </div>
      </div>
      <div tabindex="0"
           aria-selected="false"
           id="list-item-169-6"
           role="option"
           class="v-list-item v-list-item--link theme--light"
      >
        <div class="v-list-item__content">
          <div class="v-list-item__title">
            Antarctica
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
